### PR TITLE
Abstracting Controller Class

### PR DESCRIPTION
**Pull request by omangbaheti**

### CODE REVIEW

File: Assets/_Scripts/CloneController.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    private void Start()
+    {
+        StartCoroutine(WaitBeforeMerge());
+    }
+
+    
+    protected override void Movement(Vector3 dirn)
+    {
+        Debug.Log($"CanMove:{canMove}");
```
Comment (ThatAmuzak) [inline Assets/_Scripts/CloneController.cs]: Remove

File: Assets/_Scripts/CloneController.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        Vector3 targetMove = new Vector3(dirn.x * mirrorLine.x, dirn.y * mirrorLine.y, dirn.z * mirrorLine.z);
```
Comment (ThatAmuzak) [inline Assets/_Scripts/CloneController.cs]: can we not dot product?

File: Assets/_Scripts/InputHandler.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+using System.Collections;
+using UnityEngine;
+using UnityEngine.InputSystem;
+
+public class InputHandler : MonoBehaviour
+{
+    public static event Action<Vector3> MovementDirection;
+    public static event Action<Vector3> CloneCreator;
+    public static event Action ClonePreview;
+
```
Comment (ThatAmuzak) [inline Assets/_Scripts/InputHandler.cs]: I shall build my living room in this space

File: Assets/_Scripts/PlayerController.cs [hunk 4]
```diff
@@ -59,8 +64,13 @@ private void CreateClone(Vector3 movementDirn)
...
 
-        GameObject newClone=Instantiate(clone, transform.position, Quaternion.identity, cloneContainer);
-        Movement(movementDirn);
+        GameObject newClone = Instantiate(clone, transform.position, Quaternion.identity, cloneContainer);
         newClone.GetComponent<CloneController>().Spawn(transform.position, -movementDirn);
     }
+    
+    private IEnumerator cloneCooldown()
+    {
+        yield return new WaitForSeconds(0.2f);
```
Comment (ThatAmuzak) [inline Assets/_Scripts/PlayerController.cs]: sfield this as well no?

File: Assets/_Scripts/Tile.cs [hunk 5]
```diff
@@ -5,11 +5,11 @@
 
 public class Tile : MonoBehaviour
 {
-    private List<Vector3> validDirections = new();
+    private List<Vector3> validDirections = new List<Vector3>();
     private Vector3[] checklist;
     private float destroyTime = 0.3f;
 
-    private void Start()
+    private void Awake()
```
Comment (ThatAmuzak) [inline Assets/_Scripts/Tile.cs:12]: Not that I'm asking this be changed, but why awake and not start?

**Review by ThatAmuzak — CHANGES_REQUESTED** [review]
There is one thing I noted about the input actions, which is that certain functions are being triggered OnKeyDown as well as Up. Not sure if this will ever be a real problem, but maybe check the input actions asset just in case

File: Assets/_Scripts/Tile.cs (hunk 5 above)
Comment (omangbaheti) [reply Assets/_Scripts/Tile.cs:12]: checkList gives a null reference at the very start once becuase player controller calls GetValidDirections() one of the first things in the project but CheckList is null at that specific frame. So i moved it to awake and it doesnt give those errors anymore

File: Assets/_Scripts/CloneController.cs (hunk 2 above)
Comment (omangbaheti) [reply Assets/_Scripts/CloneController.cs]: dot product returns a float

File: Assets/_Scripts/InputHandler.cs (hunk 3 above)
Comment (omangbaheti) [reply Assets/_Scripts/InputHandler.cs]: I admire that youre getting prepared for the unaffordable housing in metropolitans

File: Assets/_Scripts/CloneController.cs (hunk 2 above)
Comment (omangbaheti) [reply Assets/_Scripts/CloneController.cs]: We want to do a Vector3.Scale()

**Review by ThatAmuzak — APPROVED** [review]
gg

**Pull request merged**